### PR TITLE
[FL-3593] Picopass rework with new NFC API

### PR DESCRIPTION
**Pull request by gornekich**
# What's new

- Rework Picopass App with new NFC API

## Merge after https://github.com/flipperdevices/flipperzero-firmware/pull/3050 is merged

# Verification 

- Build SDK from https://github.com/flipperdevices/flipperzero-firmware/pull/3050 : `./fbt updater_package`
- Update local path to SDK : `ufbt update -t f7 -l <local-sdk-path>`
- Build and run Picopass App: `ufbt launch`
- Verify app is working as before

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by gornekich** [comment]
Hello @bettse, @nvx !
I reworked Picopass with new NFC API. Since I don't have many picopass cards, I kindly ask you to test reworked app and make a code review. Please, let me know if you have any problems / questions

File: picopass/scenes/picopass_scene_config.h [hunk 1]
```diff
@@ -15,7 +15,7 @@ ADD_SCENE(picopass, write_card_failure, WriteCardFailure)
 ADD_SCENE(picopass, read_factory_success, ReadFactorySuccess)
 ADD_SCENE(picopass, write_key, WriteKey)
 ADD_SCENE(picopass, key_menu, KeyMenu)
-ADD_SCENE(picopass, elite_dict_attack, EliteDictAttack)
 ADD_SCENE(picopass, emulate, Emulate)
 ADD_SCENE(picopass, loclass, Loclass)
 ADD_SCENE(picopass, key_input, KeyInput)
+ADD_SCENE(picopass, dict_attack, DictAttack)
```
Comment (bettse) [inline picopass/scenes/picopass_scene_config.h:21]: Some of these seem like changes that at independent of the NFC refactor.  If that is the case, could they be pulled out and done as an independent PR(s) so that it is easier to see the changed the refactor requires?

Comment (nvx) [reply picopass/scenes/picopass_scene_config.h:21]: +1 on this. It seems like it's a NFC refactor, plus major cleanup at the same time.

I'm also concerned that some of the decisions, especially around emulation splitting to looping over an array of handlers rather than a switch statement, will affect emulation speed breaking compatibility with some readers. There's definitely some code that doesn't look the clanest but was done that way for performance reasons to make it work. Rather difficult determine what has broken something when there's so many things done together.

File: picopass/scenes/picopass_scene_key_menu.c [hunk 2]
```diff
@@ -9,6 +9,39 @@ enum SubmenuIndex {
     SubmenuIndexWriteCustom,
 };
 
+typedef struct {
```
Comment (bettse) [inline picopass/scenes/picopass_scene_key_menu.c:12]: This is an example of code I like, but feels unrelated and could probably go in beforehand (or after, which ever is easier)

**Comment by skotopes** [comment]
> Splitting PR into 2:
> - RFAL_PICOPASS_x 
> - dictionary? (check complexity)
> - Menu?

**Comment by pcunning** [comment]
I ran this through some testing

Reading seems to be working as expected, tested cards with
- different bit lengths 
- SIO
- SE
- outside HID range

I had some issues initially saving files, not sure if that was missing directories (fresh SD card) but it eventually started working.

Emulation is working on SE (Rev E) and Signo readers. It is not working on Rev A or Rev C (multiclass non SE) while the older firmware is working. (I believe this has to do with iso15 vs iso14b @nvx would know better).

When running loclass on a standard keyed reader it should indicate std keyed however it is crashing (with the vibration running) and needs a hard reset. Loclass was working as expected on an elite keyed reader.

**Comment by gornekich** [comment]
@bettse @nvx 
I will split the PR into two parts as we discussed. The first part will be refactored "worker" part with new API and the second part will be application clean up. Coming later this week

**Comment by nvx** [comment]
> Emulation is working on SE (Rev E) and Signo readers. It is not working on Rev A or Rev C (multiclass non SE) while the older firmware is working. (I believe this has to do with iso15 vs iso14b @nvx would know better).

They both do 15693, but the RevA though C readers use shallow modulation/10% modulation index ASK, while the rest you tested that worked use OOK modulation.

I haven't looked at the NFC refactor code yet, but if someone stopped using the IRQ pin and instead changed to MISO to read the incoming modulation from the card in the 15693 code that would explain the issue.

Otherwise it could be a timing issue as well, each firmware seems to have slightly different thresholds for timing go figure.

**Comment by bettse** [comment]
In an attempt to be helpful I created a PR that renamed RFAL_PICOPASS_BLOCK_LEN to make this PR smaller: https://github.com/flipperdevices/flipperzero-good-faps/pull/63
If that sounds OK we can merge it, but if you'd prefer to handle it yourself we can just close it.

**Comment by bettse** [comment]
Want me to fix the conflicting files (since they are pretty much my fault)?

**Comment by gornekich** [comment]
> Want me to fix the conflicting files (since they are pretty much my fault)?

No, thanks!
I will close this PR once I make another one with smaller changes. I leave this PR opened for a few days in case someone wants to test picopass on new NFC stack. We already fixed one issue in NFC library after testing this PR.

**Comment by skotopes** [comment]
Hello everyone, FW is almost ready: there will be one more tests iteration by QA team on Monday and then it will be released.

File: picopass/scenes/picopass_scene_config.h (hunk 1 above)
Comment (gornekich) [reply picopass/scenes/picopass_scene_config.h:21]: @nvx sorry for late response, lost your message.

I think that introduced array of handlers in emulation part doesn't affect flipper response time much but gives cleaner code. Meanwhile we improved ISO15693 signal decoder and encoder in NFC library, which helped us to achieve better timings. Still we need to test it, and if it doesn't work good, we can apply some optimizations.

**Comment by gornekich** [comment]
As we agreed, I made another PR with smaller changes, which works with new NFC stack. Let's continue conversation there #68
